### ISSUE TITLE
Datastore - observeQuery() throws exception when more than one Auth (multiauth) is configured in `amplifyconffiguration.dart`

### ISSUE THREAD

**Issue by dorontal** [aws-amplify/amplify-flutter#1867]
### Description

When you have more than one Auth configured in `amplifyconfiguration.dart` - the function `Amplify.DataStore.observeQuery()` throws an exception.  You can see in the console errors that are provided from the Android emulator that the failure is due to the fact that Amplify Flutter cannot yet handle multiple Auth in the same configuration file. But the Amplify Flutter documentation explains in detail that this should be available, (see https://docs.amplify.aws/lib/datastore/setup-auth-rules/q/platform/flutter/#configure-multiple-authorization-types).

Here is the complaint that I see in the VSCode console - which comes from the Android emulated OS:
```
I/amplify:aws-datastore( 4037): Orchestrator lock acquired.
I/amplify:aws-datastore( 4037): Orchestrator transitioning from LOCAL_ONLY to SYNC_VIA_API
I/amplify:aws-datastore( 4037): Setting currentState to SYNC_VIA_API
I/amplify:aws-datastore( 4037): Starting API synchronization mode.
I/amplify:aws-datastore( 4037): Starting processing subscription events.
I/amplify:aws-datastore( 4037): Orchestrator lock released.
E/amplify:aws-datastore( 4037): Failure encountered while attempting to start API sync.
E/amplify:aws-datastore( 4037): DataStoreException{message=Error during subscription., cause=ApiException{message=There is more than one API configured for this plugin with matching endpoint type., cause=null, recoverySuggestion=Please specify the name of API to invoke in the API method.}, recoverySuggestion=Evaluate details.}
E/amplify:aws-datastore( 4037): 	at com.amplifyframework.datastore.appsync.AppSyncClient.lambda$subscription$3(AppSyncClient.java:331)
E/amplify:aws-datastore( 4037): 	at com.amplifyframework.datastore.appsync.-$$Lambda$AppSyncClient$797ziDK0io-qXODzROLOA77stS8.accept(Unknown Source:4)
E/amplify:aws-datastore( 4037): 	at com.amplifyframework.api.aws.AWSApiPlugin.subscribe(AWSApiPlugin.java:285)
E/amplify:aws-datastore( 4037): 	at com.amplifyframework.api.ApiCategory.subscribe(ApiCategory.java:91)
E/amplify:aws-datastore( 4037): 	at com.amplifyframework.datastore.appsync.AppSyncClient.subscription(AppSyncClient.java:335)
E/amplify:aws-datastore( 4037): 	at com.amplifyframework.datastore.appsync.AppSyncClient.onCreate(AppSyncClient.java:257)
E/amplify:aws-datastore( 4037): 	at com.amplifyframework.datastore.syncengine.-$$Lambda$scUANamNqt4NIIofrmWYQsBqVJM.subscribe(Unknown Source:7)
E/amplify:aws-datastore( 4037): 	at com.amplifyframework.datastore.syncengine.SubscriptionProcessor.lambda$subscriptionObservable$6$SubscriptionProcessor(SubscriptionProcessor.java:187)
E/amplify:aws-datastore( 4037): 	at com.amplifyframework.datastore.syncengine.-$$Lambda$SubscriptionProcessor$w6tohapLGUGmW4mOmsvNOno7GVE.subscribe(Unknown Source:11)
E/amplify:aws-datastore( 4037): 	at io.reactivex.rxjava3.internal.operators.observable.ObservableCreate.subscribeActual(ObservableCreate.java:40)
E/amplify:aws-datastore( 4037): 	at io.reactivex.rxjava3.core.Observable.subscribe(Observable.java:13099)
E/amplify:aws-datastore( 4037): 	at io.reactivex.rxjava3.internal.operators.observable.ObservableDoOnEach.subscribeActual(ObservableDoOnEach.java:42)
E/amplify:aws-datastore( 4037): 	at io.reactivex.rxjava3.core.Observable.subscribe(Observable.java:13099)
E/amplify:aws-datastore( 4037): 	at io.reactivex.rxjava3.internal.operators.observable.ObservableSubscribeOn$SubscribeTask.run(ObservableSubscribeOn.java:96)
E/amplify:aws-datastore( 4037): 	at io.reactivex.rxjava3.core.Scheduler$DisposeTask.run(Scheduler.java:614)
E/amplify:aws-datastore( 4037): 	at io.reactivex.rxjava3.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:65)
E/amplify:aws-datastore( 4037): 	at io.reactivex.rxjava3.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:56)
E/amplify:aws-datastore( 4037): 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
E/amplify:aws-datastore( 4037): 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)
E/amplify:aws-datastore( 4037): 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
E/amplify:aws-datastore( 4037): 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
E/amplify:aws-datastore( 4037): 	at java.lang.Thread.run(Thread.java:764)
E/amplify:aws-datastore( 4037): Caused by: ApiException{message=There is more than one API configured for this plugin with matching endpoint type., cause=null, recoverySuggestion=Please specify the name of API to invoke in the API method.}
E/amplify:aws-datastore( 4037): 	at com.amplifyframework.api.aws.AWSApiPlugin.selectApiName(AWSApiPlugin.java:579)
E/amplify:aws-datastore( 4037): 	at com.amplifyframework.api.aws.AWSApiPlugin.getSelectedApiName(AWSApiPlugin.java:564)
E/amplify:aws-datastore( 4037): 	at com.amplifyframework.api.aws.AWSApiPlugin.subscribe(AWSApiPlugin.java:283)
E/amplify:aws-datastore( 4037): 	... 19 more
W/amplify:aws-datastore( 4037): API sync failed - transitioning to LOCAL_ONLY.
W/amplify:aws-datastore( 4037): DataStoreException{message=Error during subscription., cause=ApiException{message=There is more than one API configured for this plugin with matching endpoint type., cause=null, recoverySuggestion=Please specify the name of API to invoke in the API method.}, recoverySuggestion=Evaluate details.}
W/amplify:aws-datastore( 4037): 	at com.amplifyframework.datastore.appsync.AppSyncClient.lambda$subscription$3(AppSyncClient.java:331)
W/amplify:aws-datastore( 4037): 	at com.amplifyframework.datastore.appsync.-$$Lambda$AppSyncClient$797ziDK0io-qXODzROLOA77stS8.accept(Unknown Source:4)
W/amplify:aws-datastore( 4037): 	at com.amplifyframework.api.aws.AWSApiPlugin.subscribe(AWSApiPlugin.java:285)
W/amplify:aws-datastore( 4037): 	at com.amplifyframework.api.ApiCategory.subscribe(ApiCategory.java:91)
W/amplify:aws-datastore( 4037): 	at com.amplifyframework.datastore.appsync.AppSyncClient.subscription(AppSyncClient.java:335)
W/amplify:aws-datastore( 4037): 	at com.amplifyframework.datastore.appsync.AppSyncClient.onCreate(AppSyncClient.java:257)
W/amplify:aws-datastore( 4037): 	at com.amplifyframework.datastore.syncengine.-$$Lambda$scUANamNqt4NIIofrmWYQsBqVJM.subscribe(Unknown Source:7)
W/amplify:aws-datastore( 4037): 	at com.amplifyframework.datastore.syncengine.SubscriptionProcessor.lambda$subscriptionObservable$6$SubscriptionProcessor(SubscriptionProcessor.java:187)
W/amplify:aws-datastore( 4037): 	at com.amplifyframework.datastore.syncengine.-$$Lambda$SubscriptionProcessor$w6tohapLGUGmW4mOmsvNOno7GVE.subscribe(Unknown Source:11)
W/amplify:aws-datastore( 4037): 	at io.reactivex.rxjava3.internal.operators.observable.ObservableCreate.subscribeActual(ObservableCreate.java:40)
W/amplify:aws-datastore( 4037): 	at io.reactivex.rxjava3.core.Observable.subscribe(Observable.java:13099)
W/amplify:aws-datastore( 4037): 	at io.reactivex.rxjava3.internal.operators.observable.ObservableDoOnEach.subscribeActual(ObservableDoOnEach.java:42)
W/amplify:aws-datastore( 4037): 	at io.reactivex.rxjava3.core.Observable.subscribe(Observable.java:13099)
W/amplify:aws-datastore( 4037): 	at io.reactivex.rxjava3.internal.operators.observable.ObservableSubscribeOn$SubscribeTask.run(ObservableSubscribeOn.java:96)
W/amplify:aws-datastore( 4037): 	at io.reactivex.rxjava3.core.Scheduler$DisposeTask.run(Scheduler.java:614)
W/amplify:aws-datastore( 4037): 	at io.reactivex.rxjava3.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:65)
W/amplify:aws-datastore( 4037): 	at io.reactivex.rxjava3.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:56)
W/amplify:aws-datastore( 4037): 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
W/amplify:aws-datastore( 4037): 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)
W/amplify:aws-datastore( 4037): 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
W/amplify:aws-datastore( 4037): 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
W/amplify:aws-datastore( 4037): 	at java.lang.Thread.run(Thread.java:764)
W/amplify:aws-datastore( 4037): Caused by: ApiException{message=There is more than one API configured for this plugin with matching endpoint type., cause=null, recoverySuggestion=Please specify the name of API to invoke in the API method.}
W/amplify:aws-datastore( 4037): 	at com.amplifyframework.api.aws.AWSApiPlugin.selectApiName(AWSApiPlugin.java:579)
W/amplify:aws-datastore( 4037): 	at com.amplifyframework.api.aws.AWSApiPlugin.getSelectedApiName(AWSApiPlugin.java:564)
W/amplify:aws-datastore( 4037): 	at com.amplifyframework.api.aws.AWSApiPlugin.subscribe(AWSApiPlugin.java:283)
W/amplify:aws-datastore( 4037): 	... 19 more
I/amplify:aws-datastore( 4037): Orchestrator transitioning from SYNC_VIA_API to LOCAL_ONLY
I/amplify:aws-datastore( 4037): Setting currentState to LOCAL_ONLY
I/amplify:aws-datastore( 4037): Stopping subscription processor.
I/amplify:aws-datastore( 4037): Stopped subscription processor.
```

### Categories

- [ ] Analytics
- [ ] API (REST)
- [ ] API (GraphQL)
- [X] Auth
- [ ] Authenticator
- [X] DataStore
- [ ] Storage

### Steps to Reproduce

1. Create a GraphQL API with a simple model
2. Modify `amplifyconfiguration.dart` to add a Cognito API configuration -- see how it's done in the `tracktunesCognito` section in `amplifyconfiguration.dart` below
3. Add a call to `Amplify.Datastore.observeQuery(<MODEL>)` at the start of your program and you'll see it crashing (check log lines in VSCode of the Android emulator, giving out the Android/Java-level exception details) 

### Platforms

- [ ] iOS
- [X] Android

### Android Device/Emulator API Level

API 29

### Environment

```bash
> flutter doctor
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 3.0.4, on Debian GNU/Linux 11 (bullseye) 5.10.0-16-amd64, locale en_US.UTF-8)
[✓] Android toolchain - develop for Android devices (Android SDK version 32.1.0-rc1)
[✓] Chrome - develop for the web
[✓] Linux toolchain - develop for Linux desktop
[✓] Android Studio (version 2021.1)
[✓] VS Code (version 1.69.0)
[✓] Connected device (3 available)
[✓] HTTP Host Availability

• No issues found!
```

### Dependencies

```bash
> flutter pub deps --no-dev --style=compact
Dart SDK 2.17.5
Flutter SDK 3.0.4
tracktunes_app_aws 1.0.0+1

dependencies:
- amplify_api 0.6.1 [amplify_api_android amplify_api_ios amplify_core amplify_flutter aws_common collection flutter meta plugin_platform_interface]
- amplify_auth_cognito 0.6.1 [amplify_auth_cognito_android amplify_auth_cognito_ios amplify_core aws_common collection flutter meta plugin_platform_interface]
- amplify_datastore 0.6.1 [flutter amplify_datastore_plugin_interface amplify_core plugin_platform_interface meta collection async]
- amplify_flutter 0.6.1 [amplify_core amplify_datastore_plugin_interface amplify_flutter_android amplify_flutter_ios aws_common collection flutter meta plugin_platform_interface]
- amplify_storage_s3 0.6.1 [amplify_storage_s3_android amplify_storage_s3_ios amplify_core aws_common flutter meta plugin_platform_interface]
- cupertino_icons 1.0.5
- flutter 0.0.0 [characters collection material_color_utilities meta vector_math sky_engine]
- flutter_hooks 0.18.5+1 [flutter]
- google_fonts 3.0.1 [flutter http path_provider crypto]
- hooks_riverpod 1.0.4 [collection flutter flutter_hooks flutter_riverpod riverpod state_notifier]
- image 3.2.0 [archive meta xml]
- image_picker 0.8.5+3 [flutter image_picker_android image_picker_for_web image_picker_ios image_picker_platform_interface]
- path_provider 2.0.11 [flutter path_provider_android path_provider_ios path_provider_linux path_provider_macos path_provider_platform_interface path_provider_windows]
- rflutter_alert 2.0.4 [flutter]

transitive dependencies:
- amplify_api_android 0.6.1 [flutter]
- amplify_api_ios 0.6.1 [amplify_core flutter]
- amplify_auth_cognito_android 0.6.1 [flutter]
- amplify_auth_cognito_ios 0.6.1 [amplify_core flutter]
- amplify_core 0.6.1 [aws_common collection flutter intl json_annotation meta plugin_platform_interface uuid]
- amplify_datastore_plugin_interface 0.6.1 [amplify_core collection flutter meta]
- amplify_flutter_android 0.6.1 [flutter]
- amplify_flutter_ios 0.6.1 [amplify_core flutter]
- amplify_storage_s3_android 0.6.1 [flutter]
- amplify_storage_s3_ios 0.6.1 [flutter]
- archive 3.3.0 [crypto path]
- async 2.8.2 [collection meta]
- aws_common 0.1.1 [async collection http meta stream_transform uuid]
- characters 1.2.0
- charcode 1.3.1
- clock 1.1.0
- collection 1.16.0
- cross_file 0.3.3+1 [js meta]
- crypto 3.0.2 [typed_data]
- ffi 2.0.1
- file 6.1.2 [meta path]
- flutter_plugin_android_lifecycle 2.0.6 [flutter]
- flutter_riverpod 1.0.4 [collection flutter meta riverpod state_notifier]
- flutter_web_plugins 0.0.0 [flutter js characters collection material_color_utilities meta vector_math]
- http 0.13.4 [async http_parser meta path]
- http_parser 4.0.1 [collection source_span string_scanner typed_data]
- image_picker_android 0.8.5+1 [flutter flutter_plugin_android_lifecycle image_picker_platform_interface]
- image_picker_for_web 2.1.8 [flutter flutter_web_plugins image_picker_platform_interface]
- image_picker_ios 0.8.5+5 [flutter image_picker_platform_interface]
- image_picker_platform_interface 2.5.0 [cross_file flutter http plugin_platform_interface]
- intl 0.17.0 [clock path]
- js 0.6.4
- json_annotation 4.5.0 [meta]
- material_color_utilities 0.1.4
- meta 1.7.0
- path 1.8.1
- path_provider_android 2.0.16 [flutter path_provider_platform_interface]
- path_provider_ios 2.0.10 [flutter path_provider_platform_interface]
- path_provider_linux 2.1.7 [ffi flutter path path_provider_platform_interface xdg_directories]
- path_provider_macos 2.0.6 [flutter path_provider_platform_interface]
- path_provider_platform_interface 2.0.4 [flutter platform plugin_platform_interface]
- path_provider_windows 2.1.0 [ffi flutter path path_provider_platform_interface win32]
- petitparser 5.0.0 [meta]
- platform 3.1.0
- plugin_platform_interface 2.1.2 [meta]
- process 4.2.4 [file path platform]
- riverpod 1.0.3 [collection meta state_notifier]
- sky_engine 0.0.99
- source_span 1.8.2 [collection path term_glyph]
- state_notifier 0.7.2+1 [meta]
- stream_transform 2.0.0
- string_scanner 1.1.0 [charcode source_span]
- term_glyph 1.2.0
- typed_data 1.3.1 [collection]
- uuid 3.0.6 [crypto]
- vector_math 2.1.2
- win32 2.7.0 [ffi]
- xdg_directories 0.2.0+1 [meta path process]
- xml 6.1.0 [collection meta petitparser]
```

### Device

Samsung Galaxy S7 -- emulated

### OS

Android 8

### CLI Version

9.1.0

### Additional Context

Here's the `amplifyconfiguration.dart` file:
```dart
const amplifyconfig = ''' {
    "UserAgent": "aws-amplify-cli/2.0",
    "Version": "1.0",
    "api": {
        "plugins": {
            "awsAPIPlugin": {
                "tracktunes": {
                    "endpointType": "GraphQL",
                    "endpoint": "https://xxxxxxxxxxxxxxxxxx",
                    "region": "us-east-1",
                    "authorizationType": "AWS_IAM"
                },
                "tracktunesCognito": {
                    "endpointType": "GraphQL",
                    "endpoint": "https://xxxxxxxxxxxxxxxxxx",
                    "region": "us-east-1",
                    "authorizationType": "AMAZON_COGNITO_USER_POOLS"
                },
                "lambdaFunctions": {
                    "endpointType": "REST",
                    "endpoint": "https://xxxxxxxxxxxxxxxxxx",
                    "region": "us-east-1",
                    "authorizationType": "AWS_IAM"
                },
                "lambdaFunctionsCognito": {
                    "endpointType": "REST",
                    "endpoint": "https://xxxxxxxxxxxxxxxxxx",
                    "region": "us-east-1",
                    "authorizationType": "AMAZON_COGNITO_USER_POOLS"
                }
            }
        }
    },
    "auth": {
        "plugins": {
            "awsCognitoAuthPlugin": {
                "UserAgent": "aws-amplify-cli/0.1.0",
                "Version": "0.1.0",
                "IdentityManager": {
                    "Default": {}
                },
                "CredentialsProvider": {
                    "CognitoIdentity": {
                        "Default": {
                            "PoolId": "xxxxxxxxxxxxxxxxxx",
                            "Region": "us-east-1"
                        }
                    }
                },
                "CognitoUserPool": {
                    "Default": {
                        "PoolId": "xxxxxxxxxxxxxxxxxx",
                        "AppClientId": "xxxxxxxxxxxxxxxxxx",
                        "Region": "us-east-1"
                    }
                },
                "Auth": {
                    "Default": {
                        "authenticationFlowType": "USER_SRP_AUTH",
                        "socialProviders": [],
                        "usernameAttributes": [
                            "EMAIL"
                        ],
                        "signupAttributes": [
                            "EMAIL"
                        ],
                        "passwordProtectionSettings": {
                            "passwordPolicyMinLength": 8,
                            "passwordPolicyCharacters": []
                        },
                        "mfaConfiguration": "OFF",
                        "mfaTypes": [
                            "SMS"
                        ],
                        "verificationMechanisms": [
                            "EMAIL"
                        ]
                    }
                },
                "AppSync": {
                    "Default": {
                        "ApiUrl": "xxxxxxxxxxxxxxxxxx",
                        "Region": "us-east-1",
                        "AuthMode": "AWS_IAM",
                        "ClientDatabasePrefix": "tracktunes_AWS_IAM"
                    },
                    "tracktunes_AMAZON_COGNITO_USER_POOLS": {
                        "ApiUrl": "xxxxxxxxxxxxxxxxxx",
                        "Region": "us-east-1",
                        "AuthMode": "AMAZON_COGNITO_USER_POOLS",
                        "ClientDatabasePrefix": "tracktunes_AMAZON_COGNITO_USER_POOLS"
                    }
                },
                "S3TransferUtility": {
                    "Default": {
                        "Bucket": "userdata94336-dev",
                        "Region": "us-east-1"
                    }
                }
            }
        }
    },
    "storage": {
        "plugins": {
            "awsS3StoragePlugin": {
                "bucket": "userdata94336-dev",
                "region": "us-east-1",
                "defaultAccessLevel": "guest"
            }
        }
    }
}''';
```

**Comment by HuiSF** [comment]
Hello @dorontal thanks for opening this issue.

The exception is self explanatory - `There is more than one API configured for this plugin with matching endpoint type.`. This exception gets thrown when you have multiple keys under `awsAPIPlugin` and using API plugin along with DataStore. DataStore requires only 1 key to be presented under `awsAPIPlugin`.

Why do you need to modify the config file? I think [multi-auth in Amplify library](https://docs.amplify.aws/lib/datastore/setup-auth-rules/q/platform/flutter/#multiple-authorization-types-priority-order) doesn't require manual modification.

**Comment by dorontal** [comment]
Hi @HuiSF - 

Amplify apps need multi-auth for GraphQL API, for many use cases. Any use case where you need public and private access is such  use-case, I believe.

Your own documentation explains how to add, manually, a new section, see

https://docs.amplify.aws/lib/restapi/authz/q/platform/flutter/#cognito-user-pool-authorization

and see comment here:

https://github.com/aws-amplify/amplify-flutter/issues/1539#issuecomment-1151287563

-----------------------

The exception **is** self explanatory -- but this is because of a serious bug in Amplify code rather than in my program or in the way I use Amplify, IMO -- the exception is not supposed to exclude us from using multi-auth, right?  But currently the only way not to get an exception is to stop using multi-auth in the way that your documentation tells us to.  

Multi- auth is absolutely necessary for many nontrivial apps, so this appears to be an Amplify bug that has to be fixed. Please correct me if I'm wrong.

**Comment by HuiSF** [comment]
Hi @dorontal thanks for the follow up and more details.

The document listed multi-auth modes configuration for GraphQL and REST APIs is not applicable for DataStore use case at this moment. DataStore relies on single GraphQL API to back the sync engine, and [resolves valid auth method by priority](https://docs.amplify.aws/lib/datastore/setup-auth-rules/q/platform/flutter/#multiple-authorization-types-priority-order) as multi-auth resolution. I will do the following

1. Update the documents state that REST and GraphQL multi-auth modes configurations are not applicable to DataStore
2. Create a feature request to support choose which API to be used as DataStore sync engine while configuring DataStore

**Comment by dorontal** [comment]
@HuiSF - thanks very much for your response. I just noticed that I don't yet use multi-auth for the GraphQL (just needed multi-auth for the REST, thanks for noticing that I used it there). 

So will try running this code in regular (single) Auth and will report.

**Comment by dorontal** [comment]
@HuiSF thanks for those flags and the new usage!, but my current version of the library does not allow for the `apiName` flag to be added to the call to `AmplifyDataStore()` - the named parameter `apiName` is undefined for `AmplifyDatastore()` -- I did notice that there is another important named parameter (mentioned in the documentation) that must be added and that is `authModeStrategy`, as in:
```dart
        AmplifyDataStore(
          modelProvider: ModelProvider.instance,
          authModeStrategy: AuthModeStrategy.multiAuth,
        ),
```
but before filing this issue, I looked all over for the ability to specify `apiName` as you could do, for example in direct GraphQL requests - but could not find it anywhere for Datastore.

**Comment by HuiSF** [comment]
`authModeStrategy: AuthModeStrategy.multiAuth` this parameter doesn't resolve valid auth between GraphQL API configurations. It still requires 1 GraphQL endpoint to be listed in the configuration file, (so it won't fix the exception you encountered).

`apiName: 'tracktunesCognito'` will be an upcoming improvement (I renamed your issue and marked it as a feature request), please stay tuned.

**Comment by HuiSF** [comment]
I think there is misunderstanding here...

The multi-auth modes for GraphQL and REST using Amplify API plugins is conceptually different from multi-auth in DataStore.

The former - configure multiple APIs, each API is with a different auth mode, you choose which API to use in order to choose a specific auth mode.
The latter - DataStore relies on a single GraphQL API, it resolve a usable auth method (providing valid token for this GraphQL PI) from an array of choosable auth methods (listed by `@auth` directive in model schema).

Configuring multi-auth modes in API doesn't do anything for DataStore multi-auth. Hope that makes sense. 

We need to introduce `apiName` for DataStore in order to support this particular configuration of yours, which has multiple GraphQL endpoints listed.

But if you don't need multiple GraphQL API endpoints (the configuration file has only 1 GraphQL endpoint), DataStore multi-auth is already supported.

**Comment by dorontal** [comment]
This makes sense - that APIs are allowed to use multi Auth already. That's what my app needs. The API section of `amplifyconfiguration.dart` is the only place whereI made the multi-auth change, i.e, API is the only section of the config where I added a Cognito API configuration in addition to the CLI-generated IAM API configuration. 

I did not touch any other part of the config file, only the API section.

I thought that when you use GraphQL API with mult Auth in this way, it interferes with Datastore. I thought that the `Datastore` initialization routine threw the exception when it saw that in the API section of the config there was more than one Auth - I thought the `Datastore` init routine was incompatible with these manal config changes in the API section and that this is why it threw the exception.  

Moreover, When I remove one of the two Auth sections from the config API seciton, `Datastore` no longer throws this exception. 

I think that my options now are:
1) do not use `Datastore` at al until it can tolerate the multi-Auth API config changes
2) do not use GraphQL in multi Auth mode, only use it with one Auth, then I can use `Datastore`

Unless I misunderstood, I have to wait on Datastore, for now, because my app requires multi-Auth GraphQL API. No problem!

**Comment by dorontal** [comment]
I think I see the misunderstanding: if I ever said "endpoints" I meant "configurations" - there are two configurations of GraphQL, in `amplifyconfiguration.dart` - both use the same API endpoint. 

One uses an `authorizationType` of `AWS_IAM`, while the other uses an `authorizationType` of `AMAZON_COGNITO_USER_POOLS`.

Maybe I wasn't clear but it is those two configurations (or whatever they are called) that -- whenever there is more than one such configuration - cause the exception in DataStore initialization. Thanks for the clarification.

**Comment by HuiSF** [comment]
Here's a summary of my findings:

1. API plugin (use without DataStore plugin) and DataStore plugin both support multi-auth but they are working in different ways
2. API plugin multi-auth modes is not configurable by Amplify CLI, it requires manual configuration following [this doc](https://docs.amplify.aws/lib/graphqlapi/authz/q/platform/ios/#configure-multiple-authorization-modes). Auth modes info is listed under the `api` key in configuration file. (configuration files looks like the one pasted in this issue)
3. To enable multi-auth in DataStore, a developer needs to configure multi-auth using Amplify CLI, follow the steps shown in the following screenshot, auth modes info is listed under the `auth` key

![image](https://user-images.githubusercontent.com/10602282/178595769-61a03ebc-b8bc-4601-9631-688854a566b7.png)

<details>
<summary>Expand to see the configuration file of DataStore multi-auth</summary>

```dart
const amplifyconfig = ''' {
    "UserAgent": "aws-amplify-cli/2.0",
    "Version": "1.0",
    "api": {
        "plugins": {
            "awsAPIPlugin": {
                "oncalltriagingca": {
                    "endpointType": "GraphQL",
                    "endpoint": "https://xxx.appsync-api.us-west-2.amazonaws.com/graphql",
                    "region": "us-west-2",
                    "authorizationType": "AMAZON_COGNITO_USER_POOLS",
                    "apiKey": "xxx"
                }
            }
        }
    },
    "auth": {
        "plugins": {
            "awsCognitoAuthPlugin": {
                "UserAgent": "aws-amplify-cli/0.1.0",
                "Version": "0.1.0",
                "IdentityManager": {
                    "Default": {}
                },
                "CredentialsProvider": {
                    "CognitoIdentity": {
                        "Default": {
                            "PoolId": "us-west-2:25e89d87-1f92-4b64-9b55-xxx",
                            "Region": "us-west-2"
                        }
                    }
                },
                "CognitoUserPool": {
                    "Default": {
                        "PoolId": "us-west-xxxx",
                        "AppClientId": "xxx",
                        "Region": "us-west-2"
                    }
                },
                "Auth": {
                    "Default": {
                        "authenticationFlowType": "CUSTOM_AUTH",
                        "socialProviders": [],
                        "usernameAttributes": [],
                        "signupAttributes": [
                            "EMAIL"
                        ],
                        "passwordProtectionSettings": {
                            "passwordPolicyMinLength": 8,
                            "passwordPolicyCharacters": []
                        },
                        "mfaConfiguration": "OFF",
                        "mfaTypes": [
                            "SMS"
                        ],
                        "verificationMechanisms": [
                            "EMAIL"
                        ]
                    }
                },
                "AppSync": {
                    "Default": {
                        "ApiUrl": "https://xxx.appsync-api.us-west-2.amazonaws.com/graphql",
                        "Region": "us-west-2",
                        "AuthMode": "AMAZON_COGNITO_USER_POOLS",
                        "ClientDatabasePrefix": "oncalltriagingca_AMAZON_COGNITO_USER_POOLS"
                    },
                    "oncalltriagingca_API_KEY": {
                        "ApiUrl": "https://xxx.appsync-api.us-west-2.amazonaws.com/graphql",
                        "Region": "us-west-2",
                        "AuthMode": "API_KEY",
                        "ApiKey": "da2-xb2eywexpndhdobpzhlimp2mli",
                        "ClientDatabasePrefix": "oncalltriagingca_API_KEY"
                    },
                    "oncalltriagingca_AWS_IAM": {
                        "ApiUrl": "https://xxx.appsync-api.us-west-2.amazonaws.com/graphql",
                        "Region": "us-west-2",
                        "AuthMode": "AWS_IAM",
                        "ClientDatabasePrefix": "oncalltriagingca_AWS_IAM"
                    }
                }
            }
        }
    }
}''';
```
</details>

4. If a developer enables multi-auth modes for API plugin following steps 2, it breaks DataStore functionality when DataStore plugin is used in the same App, as DataStore currently doesn’t support multiple GraphQL endpoints to be listed in the configuration file under the `api` key in the configuration file.


Hi @dorontal for the original issue you posted, it was due to point 4.

Do you want to use multi-auth with API plugin, and also need DataStore in the same App? If so, unfortunately it's not working due to the reason explained above, we would need to provide the feature from DataStore allowing specifying which GraphQL endpoints to let DataStore use.

Or do you want to use multi-auth for only DataStore? If so, please follow the Amplify CLI multi-auth configuration mentioned in point 3.

**Comment by dorontal** [comment]
Hi @HuiSF - and thank you so much for your response and for the informative explanation.

It's not that I want to use multi-auth with API plugin , my app is already using multi-auth with API plugin. It has to use multi-Auth, because it provides a services to people who don't even have an account and have never logged in (guests) like the ability to see a limited view of any user's profile or the ability to view publicly posted documents. But the GraphQL API also needs to provide many services to signed-in users, such as seeing private documents. So both API configurations are set up - one to use IAM and the other to use Cognito, as prescribed by the documentation and as you can see in `amplifyconfiguration.dart` and it works just fine.

One minor clarification of point #4 above: it's not that GraphQL uses multiple endpoints - there is always only one endpoint for GraphQL: if you look in my `amplifyconfiguration.dart` above, for example, you'll see that the URL for both `tracktunes` and `tracktunesCognito` is exactly the same endpoint.  I believe you meant "DataStore currently doesn't support multiple GraphQL configuurations or "API configurations" but it's not an endpoint, which is a URL, which is the same for both. One of the main features of GraphQL is that it always uses only one endpoint.  I think it may be clearer to use "configuration" instead of endpoint.

So yes, the conclusion remains the same: that DataStore is not yet usable when the GraphQL API is configured with more than one authorization type. I cannot use it. 

What's interesting is that the DataStore documentation on multi Auth seems to suggest that DataStore is supposed to work out of the box in this case when the GraphQL API is configured with more than one authorization type. The documentation even goes as far as to say that DataStore can pick the proper `@auth` rule automatically - and this means it knows how to pick either the IAM version or the Cognito version of the GraphQL API automatically.   That's what the documentation claims. But when I run it, it throws an exception that contradicts the documentation - I think that exception should not be thrown and instead it would be nice to see DataStore automatically pick the right GraphQL API auth configuration, as the documentation claims it can.

**Comment by sumchans** [comment]
Does anybody have any idea when this issue will be fixed?

**Comment by dorontal** [comment]
Considering how this issue has been responded to, we may have to wait ... for a very long time.

This issue was originally labeled a bug, as it needs to be, because the missing functionality reported by this issue -- multi-auth + datastore working together -- **is supported and working in the core Amplify libraries**, as well as in Amplify-Js or Amplify-React or Amplify-iOS. 

It's only in Amplify-Flutter where using multi-auth does not work together with Datastore.

But this issue has been labeled "feature request" for some reason. I do know that "feature request" labels take a whole lot longer to "fix" than bugs, on average.

This issue needs to be correctly labeled as a bug and then it needs to be fixed in the Amplify-Flutter wrapper of Datastore.

**Comment by dorontal** [comment]
Here is the documentation part that tells us how to use mult-auth with Datastore, in Amplify-Flutter: 
https://docs.amplify.aws/lib/datastore/setup-auth-rules/q/platform/flutter/#multiple-authorization-types-priority-order

This implies that Datastore is supposed to work with Multi-Auth - you even tell us the recipe of how to do it.

Why, then, is this bug here, this issue, renamed to "feature request"?  The feature is already implemented, according to your documentation, even the method whereby Datastore chooses the current Auth out of multiple Auth configurations is described in your documentation. This is not a feature request, it is documented that the feature is already implemented (for many months now).

I am disappointed with the way this issue has been handled or interpreted.

Datastore is useles without handling Multi-Auth IMO - for how long do you intend to keep it in this useless state?

**Comment by dorontal** [comment]
This

https://docs.amplify.aws/lib/datastore/setup-auth-rules/q/platform/flutter/#multiple-authorization-types-priority-order

implies that Datastore handles multi-auth in Amplify.

Please help make this happen in Amplify Flutter - it does not seem like such a big deal to add this feature and -- since multi-Auth is used a lot, for apps where guest access is needed via IAM in addition to Cognito access -- this feature seems necessary to enable a whole lot of apps to use DataStore at all.

**Comment by dorontal** [comment]
Hi @HuiSF - what is your current ETA for resolving this issue?  It has been around for almost six months - still unable to use DataStore because of it. Thanks.

**Comment by HuiSF** [comment]
Hi @dorontal sorry for the delayed response and thanks for your patience, we are communicating with amplify-swift and amplify-android maintainers as well as relevant internal teams regarding this issue to determine a suitable long term solution. I'll update the progress.

**Comment by dorontal** [comment]
Hi @HuiSF -- (I mistakenly asked this question in another issue first but have since been able to close that solved issue. Please ignore any prior comments there if you saw them) -- any info about when you intend to deal with this issue would help us do some planning over here, only if you know, thanks.

**Comment by dorontal** [comment]
@HuiSF  -- In case you were wondering why I believe multi-mode authorization is so common / necessary, here's our use case that demands it and which I believe is shared by many apps:

I'm putting this use-case here also because maybe somebody can suggest a way to implement this use-case **without** using multi-auth, which would also allow us to move forward.

One use case that can benefit from multi-auth + datastore together is this: 

1) DynamoDB saves a user-profile model for each user of our app, accessed via GraphQL
2) Some fields in the user model (e.g. how much money they've made) are private and can only be used by the owner user
3) We still want to be able to show the non-private fields of any user's profile to anyone -- even people without an account at all, or users who are not signed in, should be able to see limited user profile information about other users (e.g, if you go to your Github home page while signed out, you still can see some stats about your username there) 

If we only needed parts 1 and 2 above, we'd be able to get by with Cognito authorization alone, no multi-auth needed.

But if we want to share info about users with anybody, as in part 3, we must use **IAM** for that, hence we must use multi-auth for that.

This use case is why I think this issue is very important not just to me, but to many others. It seems to me like it would be a common use case.

**Comment by dorontal** [comment]
BTW, for those of you who need a workaround, I am using this workaround: created a lambda function that performs the needed DB operation (just a GET, in my case); the lambda function is accessed by guest users via IAM and a REST API.
